File: src/engine/FootballMetrics.ts
/**
 * GMN-Football-3 — Traceable Football Metrics Engine
 * Single authoritative source of truth for measuring football outcomes, possession,
 * passing, shooting, ball progression, defensive actions, and behavioral distributions.
 * 
 * Every metric records explicit numerators and denominators for complete scientific auditability.
 */

import { GameEngine } from './GameEngine';
import { ActionType, AgentAction, TeamSide } from '../types/football';
import { ACTION_SPACE_SIZE } from './Contract';

export interface EpisodeMetrics {
  episodeIndex: number;
  seed: number;
  scenario: string;
  durationSeconds: number;
  durationTicks: number;
  
  // Outcome metrics
  goalsScored: number;
  goalsConceded: number;
  goalDifference: number;
  isWin: boolean;
  isDraw: boolean;
  isLoss: boolean;
  isSuccess: boolean; // For scenarios with custom objectives or goal > 0

  // Possession metrics
  possessionTicksLeft: number;
  possessionTicksRight: number;
  possessionRateLeft: number; // possessionTicksLeft / max(1, totalTicks)
  timeToFirstPossessionSeconds: number | null;
  possessionChanges: number;

  // Passing metrics
  passesAttempted: number;
  passesCompleted: number;
  passCompletionRate: number; // completed / max(1, attempted)

  // Shooting metrics
  shotsTotal: number;
  shotsOnTarget: number;
  shotAccuracy: number; // onTarget / max(1, shotsTotal)
  goalsFromShots: number;

  // Ball progression metrics
  initialBallDistanceToOppGoal: number;
  finalBallDistanceToOppGoal: number;
  minBallDistanceToOppGoal: number;
  forwardProgressTotal: number; // delta toward opponent goal line (x = 1.0)
  maxBallProgressX: number;

  // Defensive metrics
  tacklesAttempted: number;
  tacklesSuccessful: number;
  interceptions: number;
  foulsCommitted: number;
  yellowCards: number;
  redCards: number;
  turnoversForced: number;
  turnoversConceded: number;

  // Behavioral action distributions
  actionCounts: number[]; // 19 discrete actions
  actionFrequencies: number[];
  movementDirectionCounts: number[]; // 8 movement directions (actions 1..8)
  idleActionRatio: number;
  passRatio: number;
  shotRatio: number;
  dribbleRatio: number;
  sprintRatio: number;

  // Rewards
  cumulativeReward: number;
  meanStepReward: number;
}

export interface MetricDistribution {
  mean: number;
  std: number;
  median: number;
  min: number;
  max: number;
  ci95Low: number;
  ci95High: number;
  numeratorTotal?: number;
  denominatorTotal?: number;
}

export interface AggregatedMetrics {
  totalEpisodes: number;
  totalSteps: number;
  scenario: string;
  policyName: string;
  seeds: number[];

  winRatePct: MetricDistribution;
  drawRatePct: MetricDistribution;
  lossRatePct: MetricDistribution;
  successRatePct: MetricDistribution;
  
  goalsScoredPerEpisode: MetricDistribution;
  goalsConcededPerEpisode: MetricDistribution;
  goalDifferencePerEpisode: MetricDistribution;

  possessionRatePct: MetricDistribution;
  timeToFirstPossessionSec: MetricDistribution;
  possessionChangesPerEpisode: MetricDistribution;

  passesAttemptedPerEpisode: MetricDistribution;
  passesCompletedPerEpisode: MetricDistribution;
  passCompletionRatePct: MetricDistribution;

  shotsPerEpisode: MetricDistribution;
  shotsOnTargetPerEpisode: MetricDistribution;
  shotAccuracyPct: MetricDistribution;

  forwardProgressPerEpisode: MetricDistribution;
  minDistanceToGoal: MetricDistribution;
  maxBallProgressX: MetricDistribution;

  tacklesPerEpisode: MetricDistribution;
  interceptionsPerEpisode: MetricDistribution;
  turnoversConcededPerEpisode: MetricDistribution;

  cumulativeReward: MetricDistribution;
  actionDistribution: number[]; // Normalized across all episodes
  idleActionRatioPct: MetricDistribution;
  passRatioPct: MetricDistribution;
  shotRatioPct: MetricDistribution;
  dribbleRatioPct: MetricDistribution;
  sprintRatioPct: MetricDistribution;
}

export class FootballMetricsTracker {
  private episodeMetricsList: EpisodeMetrics[] = [];
  
  // Current episode tracking
  private currentEpisodeIndex = 0;
  private currentSeed = 0;
  private currentScenario = '';
  private currentTicks = 0;
  private currentReward = 0;
  private initialBallDistToGoal = 0;
  private minBallDistToGoal = Infinity;
  private timeToFirstPossessionSeconds: number | null = null;
  private previousPossessionTeam: TeamSide | null = null;
  private possessionChanges = 0;
  private turnoversConceded = 0;
  private turnoversForced = 0;
  private actionCounts: number[] = new Array(ACTION_SPACE_SIZE).fill(0);
  private directionCounts: number[] = new Array(8).fill(0);

  public startEpisode(scenario: string, seed: number, initialEngine: GameEngine): void {
    this.currentScenario = scenario;
    this.currentSeed = seed;
    this.currentTicks = 0;
    this.currentReward = 0;
    this.actionCounts.fill(0);
    this.directionCounts.fill(0);
    this.possessionChanges = 0;
    this.turnoversConceded = 0;
    this.turnoversForced = 0;
    this.timeToFirstPossessionSeconds = null;
    this.previousPossessionTeam = null;

    const ballPos = initialEngine.ball.position;
    this.initialBallDistToGoal = Math.hypot(1.0 - ballPos.x, -ballPos.y);
    this.minBallDistToGoal = this.initialBallDistToGoal;
  }

  public recordTick(
    engine: GameEngine,
    leftActionIndices: number[],
    stepReward: number
  ): void {
    this.currentTicks++;
    this.currentReward += stepReward;

    // Record action distribution
    for (const aIdx of leftActionIndices) {
      if (aIdx >= 0 && aIdx < ACTION_SPACE_SIZE) {
        this.actionCounts[aIdx]++;
        if (aIdx >= 1 && aIdx <= 8) {
          this.directionCounts[aIdx - 1]++;
        }
      }
    }

    // Ball distance
    const ballPos = engine.ball.position;
    const currentDist = Math.hypot(1.0 - ballPos.x, -ballPos.y);
    if (currentDist < this.minBallDistToGoal) {
      this.minBallDistToGoal = currentDist;
    }

    // Possession tracking
    const currentOwner = engine.players.find((p) => p.id === engine.ball.ownerId);
    const currentTeam = currentOwner ? currentOwner.team : null;

    if (currentTeam === 'left' && this.timeToFirstPossessionSeconds === null) {
      this.timeToFirstPossessionSeconds = this.currentTicks / 60;
    }

    if (currentTeam && currentTeam !== this.previousPossessionTeam) {
      this.possessionChanges++;
      if (this.previousPossessionTeam === 'left' && currentTeam === 'right') {
        this.turnoversConceded++;
      } else if (this.previousPossessionTeam === 'right' && currentTeam === 'left') {
        this.turnoversForced++;
      }
      this.previousPossessionTeam = currentTeam;
    }
  }

  public endEpisode(finalEngine: GameEngine): EpisodeMetrics {
    const totalTicks = Math.max(1, this.currentTicks);
    const durationSeconds = totalTicks / 60;
    const stats = finalEngine.stats;
    const score = finalEngine.score;

    const goalsScored = score.left;
    const goalsConceded = score.right;
    const goalDifference = goalsScored - goalsConceded;
    const isWin = goalsScored > goalsConceded;
    const isDraw = goalsScored === goalsConceded;
    const isLoss = goalsScored < goalsConceded;
    const isSuccess = goalsScored > 0;

    const ballPos = finalEngine.ball.position;
    const finalBallDistToGoal = Math.hypot(1.0 - ballPos.x, -ballPos.y);
    const forwardProgressTotal = this.initialBallDistToGoal - finalBallDistToGoal;

    const totalActions = this.actionCounts.reduce((a, b) => a + b, 0);
    const actionFrequencies = this.actionCounts.map((c) => (totalActions > 0 ? c / totalActions : 0));

    // Calculate action ratios
    const idleCount = this.actionCounts[0]; // Action 0 is IDLE
    const passCount = this.actionCounts[9] + this.actionCounts[10] + this.actionCounts[11]; // LONG, HIGH, SHORT PASS
    const shotCount = this.actionCounts[12]; // SHOT
    const sprintCount = this.actionCounts[13]; // SPRINT
    const dribbleCount = this.actionCounts[17]; // DRIBBLE

    const passesAttempted = stats.passes.left;
    const passesCompleted = stats.completedPasses.left;
    const passCompletionRate = passesAttempted > 0 ? (passesCompleted / passesAttempted) * 100 : 0;

    const shotsTotal = stats.shots.left;
    const shotsOnTarget = stats.shotsOnTarget.left;
    const shotAccuracy = shotsTotal > 0 ? (shotsOnTarget / shotsTotal) * 100 : 0;

    const epMetrics: EpisodeMetrics = {
      episodeIndex: this.currentEpisodeIndex++,
      seed: this.currentSeed,
      scenario: this.currentScenario,
      durationSeconds,
      durationTicks: totalTicks,
      goalsScored,
      goalsConceded,
      goalDifference,
      isWin,
      isDraw,
      isLoss,
      isSuccess,
      possessionTicksLeft: stats.possession.left,
      possessionTicksRight: stats.possession.right,
      possessionRateLeft: stats.possession.left, // MatchEngine tracks possession as 0-100 percentage
      timeToFirstPossessionSeconds: this.timeToFirstPossessionSeconds ?? durationSeconds,
      possessionChanges: this.possessionChanges,
      passesAttempted,
      passesCompleted,
      passCompletionRate,
      shotsTotal,
      shotsOnTarget,
      shotAccuracy,
      goalsFromShots: goalsScored,
      initialBallDistanceToOppGoal: this.initialBallDistToGoal,
      finalBallDistanceToOppGoal: finalBallDistToGoal,
      minBallDistanceToOppGoal: this.minBallDistToGoal,
      forwardProgressTotal,
      maxBallProgressX: finalEngine.maxBallProgressX,
      tacklesAttempted: stats.tackles.left,
      tacklesSuccessful: stats.tackles.left, // All executed tackles are recorded
      interceptions: stats.interceptions.left,
      foulsCommitted: stats.fouls.left,
      yellowCards: stats.yellowCards.left,
      redCards: stats.redCards.left,
      turnoversForced: this.turnoversForced,
      turnoversConceded: this.turnoversConceded,
      actionCounts: [...this.actionCounts],
      actionFrequencies,
      movementDirectionCounts: [...this.directionCounts],
      idleActionRatio: totalActions > 0 ? (idleCount / totalActions) * 100 : 0,
      passRatio: totalActions > 0 ? (passCount / totalActions) * 100 : 0,
      shotRatio: totalActions > 0 ? (shotCount / totalActions) * 100 : 0,
      dribbleRatio: totalActions > 0 ? (dribbleCount / totalActions) * 100 : 0,
      sprintRatio: totalActions > 0 ? (sprintCount / totalActions) * 100 : 0,
      cumulativeReward: this.currentReward,
      meanStepReward: totalTicks > 0 ? this.currentReward / totalTicks : 0,
    };

    this.episodeMetricsList.push(epMetrics);
    return epMetrics;
  }

  public getEpisodes(): EpisodeMetrics[] {
    return this.episodeMetricsList;
  }

  public clear(): void {
    this.episodeMetricsList = [];
    this.currentEpisodeIndex = 0;
  }

  public aggregate(policyName: string, scenario: string): AggregatedMetrics {
    const episodes = this.episodeMetricsList;
    const n = episodes.length;
    if (n === 0) {
      throw new Error('[FootballMetricsTracker] Cannot aggregate empty episodes list.');
    }

    const seeds = Array.from(new Set(episodes.map((e) => e.seed)));
    const totalSteps = episodes.reduce((sum, e) => sum + e.durationTicks, 0);

    const winValues = episodes.map((e) => (e.isWin ? 100 : 0));
    const drawValues = episodes.map((e) => (e.isDraw ? 100 : 0));
    const lossValues = episodes.map((e) => (e.isLoss ? 100 : 0));
    const successValues = episodes.map((e) => (e.isSuccess ? 100 : 0));

    const totalPassesAttempted = episodes.reduce((s, e) => s + e.passesAttempted, 0);
    const totalPassesCompleted = episodes.reduce((s, e) => s + e.passesCompleted, 0);
    const totalShots = episodes.reduce((s, e) => s + e.shotsTotal, 0);
    const totalShotsOnTarget = episodes.reduce((s, e) => s + e.shotsOnTarget, 0);

    // Aggregated action distribution
    const combinedActionCounts = new Array(ACTION_SPACE_SIZE).fill(0);
    for (const ep of episodes) {
      for (let i = 0; i < ACTION_SPACE_SIZE; i++) {
        combinedActionCounts[i] += ep.actionCounts[i];
      }
    }
    const totalActionSum = combinedActionCounts.reduce((a, b) => a + b, 0);
    const actionDistribution = combinedActionCounts.map((c) => (totalActionSum > 0 ? c / totalActionSum : 0));

    return {
      totalEpisodes: n,
      totalSteps,
      scenario,
      policyName,
      seeds,
      winRatePct: computeDistribution(winValues, episodes.filter((e) => e.isWin).length, n),
      drawRatePct: computeDistribution(drawValues, episodes.filter((e) => e.isDraw).length, n),
      lossRatePct: computeDistribution(lossValues, episodes.filter((e) => e.isLoss).length, n),
      successRatePct: computeDistribution(successValues, episodes.filter((e) => e.isSuccess).length, n),

      goalsScoredPerEpisode: computeDistribution(episodes.map((e) => e.goalsScored)),
      goalsConcededPerEpisode: computeDistribution(episodes.map((e) => e.goalsConceded)),
      goalDifferencePerEpisode: computeDistribution(episodes.map((e) => e.goalDifference)),

      possessionRatePct: computeDistribution(episodes.map((e) => e.possessionRateLeft)),
      timeToFirstPossessionSec: computeDistribution(
        episodes.map((e) => (e.timeToFirstPossessionSeconds !== null ? e.timeToFirstPossessionSeconds : e.durationSeconds))
      ),
      possessionChangesPerEpisode: computeDistribution(episodes.map((e) => e.possessionChanges)),

      passesAttemptedPerEpisode: computeDistribution(episodes.map((e) => e.passesAttempted)),
      passesCompletedPerEpisode: computeDistribution(episodes.map((e) => e.passesCompleted)),
      passCompletionRatePct: computeDistribution(
        episodes.map((e) => e.passCompletionRate),
        totalPassesCompleted,
        totalPassesAttempted
      ),

      shotsPerEpisode: computeDistribution(episodes.map((e) => e.shotsTotal)),
      shotsOnTargetPerEpisode: computeDistribution(episodes.map((e) => e.shotsOnTarget)),
      shotAccuracyPct: computeDistribution(
        episodes.map((e) => e.shotAccuracy),
        totalShotsOnTarget,
        totalShots
      ),

      forwardProgressPerEpisode: computeDistribution(episodes.map((e) => e.forwardProgressTotal)),
      minDistanceToGoal: computeDistribution(episodes.map((e) => e.minBallDistanceToOppGoal)),
      maxBallProgressX: computeDistribution(episodes.map((e) => e.maxBallProgressX)),

      tacklesPerEpisode: computeDistribution(episodes.map((e) => e.tacklesAttempted)),
      interceptionsPerEpisode: computeDistribution(episodes.map((e) => e.interceptions)),
      turnoversConcededPerEpisode: computeDistribution(episodes.map((e) => e.turnoversConceded)),

      cumulativeReward: computeDistribution(episodes.map((e) => e.cumulativeReward)),
      actionDistribution,
      idleActionRatioPct: computeDistribution(episodes.map((e) => e.idleActionRatio)),
      passRatioPct: computeDistribution(episodes.map((e) => e.passRatio)),
      shotRatioPct: computeDistribution(episodes.map((e) => e.shotRatio)),
      dribbleRatioPct: computeDistribution(episodes.map((e) => e.dribbleRatio)),
      sprintRatioPct: computeDistribution(episodes.map((e) => e.sprintRatio)),
    };
  }
}

export function computeDistribution(
  values: number[],
  numeratorTotal?: number,
  denominatorTotal?: number
): MetricDistribution {
  const n = values.length;
  if (n === 0) {
    return { mean: 0, std: 0, median: 0, min: 0, max: 0, ci95Low: 0, ci95High: 0 };
  }

  const sum = values.reduce((a, b) => a + b, 0);
  const mean = sum / n;

  const variance = values.reduce((s, v) => s + (v - mean) ** 2, 0) / (n > 1 ? n - 1 : 1);
  const std = Math.sqrt(Math.max(0, variance));

  const sorted = [...values].sort((a, b) => a - b);
  const median = n % 2 === 0 ? (sorted[n / 2 - 1] + sorted[n / 2]) / 2 : sorted[Math.floor(n / 2)];
  const min = sorted[0];
  const max = sorted[n - 1];

  // 95% Confidence Interval for the mean
  const sem = std / Math.sqrt(n);
  const z95 = 1.96;
  const ci95Low = mean - z95 * sem;
  const ci95High = mean + z95 * sem;

  return {
    mean,
    std,
    median,
    min,
    max,
    ci95Low,
    ci95High,
    numeratorTotal,
    denominatorTotal,
  };
}

File: training/football_metrics.py
"""
GMN-Football-3 — Traceable Football Metrics Engine (Python)
Collects structured, traceable metrics across match episodes.
Calculates outcome, possession, passing, shooting, ball progression, defensive, and behavioral stats.
Records explicit numerators and denominators for complete scientific reproducibility.
"""

import math
from typing import Dict, List, Any, Optional
from dataclasses import dataclass, field, asdict


@dataclass
class EpisodeMetrics:
    episode_index: int
    seed: int
    scenario: str
    duration_seconds: float
    duration_ticks: int

    # Outcome metrics
    goals_scored: int
    goals_conceded: int
    goal_difference: int
    is_win: bool
    is_draw: bool
    is_loss: bool
    is_success: bool

    # Possession metrics
    possession_ticks_left: float
    possession_ticks_right: float
    possession_rate_left: float
    time_to_first_possession_seconds: float
    possession_changes: int

    # Passing metrics
    passes_attempted: int
    passes_completed: int
    pass_completion_rate: float

    # Shooting metrics
    shots_total: int
    shots_on_target: int
    shot_accuracy: float
    goals_from_shots: int

    # Ball progression metrics
    initial_ball_dist_to_opp_goal: float
    final_ball_dist_to_opp_goal: float
    min_ball_dist_to_opp_goal: float
    forward_progress_total: float
    max_ball_progress_x: float

    # Defensive metrics
    tackles_attempted: int
    tackles_successful: int
    interceptions: int
    fouls_committed: int
    yellow_cards: int
    red_cards: int
    turnovers_forced: int
    turnovers_conceded: int

    # Behavioral action distributions
    action_counts: List[int]
    action_frequencies: List[float]
    movement_direction_counts: List[int]
    idle_action_ratio: float
    pass_ratio: float
    shot_ratio: float
    dribble_ratio: float
    sprint_ratio: float

    # Rewards
    cumulative_reward: float
    mean_step_reward: float


def compute_distribution(
    values: List[float],
    numerator_total: Optional[int] = None,
    denominator_total: Optional[int] = None
) -> Dict[str, Any]:
    n = len(values)
    if n == 0:
        return {
            "mean": 0.0, "std": 0.0, "median": 0.0,
            "min": 0.0, "max": 0.0, "ci95_low": 0.0, "ci95_high": 0.0
        }

    mean_val = sum(values) / n
    variance = sum((v - mean_val) ** 2 for v in values) / (n - 1 if n > 1 else 1)
    std_val = math.sqrt(max(0.0, variance))

    sorted_vals = sorted(values)
    median_val = (sorted_vals[n // 2 - 1] + sorted_vals[n // 2]) / 2.0 if n % 2 == 0 else sorted_vals[n // 2]
    min_val = sorted_vals[0]
    max_val = sorted_vals[-1]

    sem = std_val / math.sqrt(n)
    ci95_low = mean_val - 1.96 * sem
    ci95_high = mean_val + 1.96 * sem

    res = {
        "mean": mean_val,
        "std": std_val,
        "median": median_val,
        "min": min_val,
        "max": max_val,
        "ci95_low": ci95_low,
        "ci95_high": ci95_high,
    }
    if numerator_total is not None:
        res["numerator_total"] = numerator_total
    if denominator_total is not None:
        res["denominator_total"] = denominator_total
    return res


class FootballMetricsTracker:
    def __init__(self):
        self.episodes: List[EpisodeMetrics] = []
        self.action_space_size = 19
        self.current_episode_index = 0
        self._reset_current_episode()

    def _reset_current_episode(self):
        self.current_scenario = ""
        self.current_seed = 0
        self.current_ticks = 0
        self.current_reward = 0.0
        self.action_counts = [0] * self.action_space_size
        self.direction_counts = [0] * 8
        self.initial_ball_dist = 0.0
        self.min_ball_dist = float("inf")
        self.time_to_first_possession = None
        self.previous_possession_team = None
        self.possession_changes = 0
        self.turnovers_conceded = 0
        self.turnovers_forced = 0

    def start_episode(self, scenario: str, seed: int, ball_pos: Dict[str, float]):
        self._reset_current_episode()
        self.current_scenario = scenario
        self.current_seed = seed
        bx = ball_pos.get("x", 0.0)
        by = ball_pos.get("y", 0.0)
        self.initial_ball_dist = math.hypot(1.0 - bx, by)
        self.min_ball_dist = self.initial_ball_dist

    def record_tick(
        self,
        left_action_indices: List[int],
        step_reward: float,
        ball_pos: Dict[str, float],
        owner_team: Optional[str]
    ):
        self.current_ticks += 1
        self.current_reward += step_reward

        for a_idx in left_action_indices:
            if 0 <= a_idx < self.action_space_size:
                self.action_counts[a_idx] += 1
                if 1 <= a_idx <= 8:
                    self.direction_counts[a_idx - 1] += 1

        bx = ball_pos.get("x", 0.0)
        by = ball_pos.get("y", 0.0)
        dist = math.hypot(1.0 - bx, by)
        if dist < self.min_ball_dist:
            self.min_ball_dist = dist

        if owner_team == "left" and self.time_to_first_possession is None:
            self.time_to_first_possession = self.current_ticks / 60.0

        if owner_team and owner_team != self.previous_possession_team:
            self.possession_changes += 1
            if self.previous_possession_team == "left" and owner_team == "right":
                self.turnovers_conceded += 1
            elif self.previous_possession_team == "right" and owner_team == "left":
                self.turnovers_forced += 1
            self.previous_possession_team = owner_team

    def end_episode(
        self,
        final_score: Dict[str, int],
        final_stats: Dict[str, Any],
        final_ball_pos: Dict[str, float],
        max_ball_progress_x: float = 0.0
    ) -> EpisodeMetrics:
        total_ticks = max(1, self.current_ticks)
        duration_seconds = total_ticks / 60.0

        goals_scored = final_score.get("left", 0)
        goals_conceded = final_score.get("right", 0)
        goal_diff = goals_scored - goalsConceded if "goalsConceded" in locals() else goals_scored - goals_conceded
        is_win = goals_scored > goals_conceded
        is_draw = goals_scored == goals_conceded
        is_loss = goals_scored < goals_conceded
        is_success = goals_scored > 0

        bx = final_ball_pos.get("x", 0.0)
        by = final_ball_pos.get("y", 0.0)
        final_dist = math.hypot(1.0 - bx, by)
        forward_progress = self.initial_ball_dist - final_dist

        total_actions = sum(self.action_counts)
        freqs = [c / total_actions if total_actions > 0 else 0.0 for c in self.action_counts]

        idle_count = self.action_counts[0]
        pass_count = sum(self.action_counts[9:12])
        shot_count = self.action_counts[12]
        sprint_count = self.action_counts[13]
        dribble_count = self.action_counts[17]

        passes_attempted = final_stats.get("passes", {}).get("left", 0) if isinstance(final_stats.get("passes"), dict) else 0
        passes_completed = final_stats.get("completedPasses", {}).get("left", 0) if isinstance(final_stats.get("completedPasses"), dict) else 0
        pass_comp_rate = (passes_completed / passes_attempted * 100.0) if passes_attempted > 0 else 0.0

        shots_total = final_stats.get("shots", {}).get("left", 0) if isinstance(final_stats.get("shots"), dict) else 0
        shots_on_target = final_stats.get("shotsOnTarget", {}).get("left", 0) if isinstance(final_stats.get("shotsOnTarget"), dict) else 0
        shot_acc = (shots_on_target / shots_total * 100.0) if shots_total > 0 else 0.0

        possession_left = final_stats.get("possession", {}).get("left", 50.0) if isinstance(final_stats.get("possession"), dict) else 50.0

        ep = EpisodeMetrics(
            episode_index=self.current_episode_index,
            seed=self.current_seed,
            scenario=self.current_scenario,
            duration_seconds=duration_seconds,
            duration_ticks=total_ticks,
            goals_scored=goals_scored,
            goals_conceded=goals_conceded,
            goal_difference=goal_diff,
            is_win=is_win,
            is_draw=is_draw,
            is_loss=is_loss,
            is_success=is_success,
            possession_ticks_left=possession_left,
            possession_ticks_right=100.0 - possession_left,
            possession_rate_left=possession_left,
            time_to_first_possession_seconds=self.time_to_first_possession if self.time_to_first_possession is not None else duration_seconds,
            possession_changes=self.possession_changes,
            passes_attempted=passes_attempted,
            passes_completed=passes_completed,
            pass_completion_rate=pass_comp_rate,
            shots_total=shots_total,
            shots_on_target=shots_on_target,
            shot_accuracy=shot_acc,
            goals_from_shots=goals_scored,
            initial_ball_dist_to_opp_goal=self.initial_ball_dist,
            final_ball_dist_to_opp_goal=final_dist,
            min_ball_dist_to_opp_goal=self.min_ball_dist,
            forward_progress_total=forward_progress,
            max_ball_progress_x=max_ball_progress_x,
            tackles_attempted=final_stats.get("tackles", {}).get("left", 0) if isinstance(final_stats.get("tackles"), dict) else 0,
            tackles_successful=final_stats.get("tackles", {}).get("left", 0) if isinstance(final_stats.get("tackles"), dict) else 0,
            interceptions=final_stats.get("interceptions", {}).get("left", 0) if isinstance(final_stats.get("interceptions"), dict) else 0,
            fouls_committed=final_stats.get("fouls", {}).get("left", 0) if isinstance(final_stats.get("fouls"), dict) else 0,
            yellow_cards=final_stats.get("yellowCards", {}).get("left", 0) if isinstance(final_stats.get("yellowCards"), dict) else 0,
            red_cards=final_stats.get("redCards", {}).get("left", 0) if isinstance(final_stats.get("redCards"), dict) else 0,
            turnovers_forced=self.turnovers_forced,
            turnovers_conceded=self.turnovers_conceded,
            action_counts=list(self.action_counts),
            action_frequencies=freqs,
            movement_direction_counts=list(self.direction_counts),
            idle_action_ratio=(idle_count / total_actions * 100.0) if total_actions > 0 else 0.0,
            pass_ratio=(pass_count / total_actions * 100.0) if total_actions > 0 else 0.0,
            shot_ratio=(shot_count / total_actions * 100.0) if total_actions > 0 else 0.0,
            dribble_ratio=(dribble_count / total_actions * 100.0) if total_actions > 0 else 0.0,
            sprint_ratio=(sprint_count / total_actions * 100.0) if total_actions > 0 else 0.0,
            cumulative_reward=self.current_reward,
            mean_step_reward=self.current_reward / total_ticks if total_ticks > 0 else 0.0,
        )

        self.current_episode_index += 1
        self.episodes.append(ep)
        return ep

    def aggregate(self, policy_name: str, scenario: str) -> Dict[str, Any]:
        episodes = self.episodes
        n = len(episodes)
        if n == 0:
            raise ValueError("[FootballMetricsTracker] No episodes to aggregate.")

        seeds = sorted(list(set(e.seed for e in episodes)))
        total_steps = sum(e.duration_ticks for e in episodes)

        win_vals = [100.0 if e.is_win else 0.0 for e in episodes]
        draw_vals = [100.0 if e.is_draw else 0.0 for e in episodes]
        loss_vals = [100.0 if e.is_loss else 0.0 for e in episodes]
        success_vals = [100.0 if e.is_success else 0.0 for e in episodes]

        tot_passes_att = sum(e.passes_attempted for e in episodes)
        tot_passes_comp = sum(e.passes_completed for e in episodes)
        tot_shots = sum(e.shots_total for e in episodes)
        tot_shots_on_target = sum(e.shots_on_target for e in episodes)

        # Action distribution
        combined_action_counts = [0] * self.action_space_size
        for ep in episodes:
            for i in range(self.action_space_size):
                combined_action_counts[i] += ep.action_counts[i]
        tot_act = sum(combined_action_counts)
        action_dist = [c / tot_act if tot_act > 0 else 0.0 for c in combined_action_counts]

        return {
            "total_episodes": n,
            "total_steps": total_steps,
            "scenario": scenario,
            "policy_name": policy_name,
            "seeds": seeds,
            "win_rate_pct": compute_distribution(win_vals, sum(1 for e in episodes if e.is_win), n),
            "draw_rate_pct": compute_distribution(draw_vals, sum(1 for e in episodes if e.is_draw), n),
            "loss_rate_pct": compute_distribution(loss_vals, sum(1 for e in episodes if e.is_loss), n),
            "success_rate_pct": compute_distribution(success_vals, sum(1 for e in episodes if e.is_success), n),

            "goals_scored_per_episode": compute_distribution([float(e.goals_scored) for e in episodes]),
            "goals_conceded_per_episode": compute_distribution([float(e.goals_conceded) for e in episodes]),
            "goal_difference_per_episode": compute_distribution([float(e.goal_difference) for e in episodes]),

            "possession_rate_pct": compute_distribution([e.possession_rate_left for e in episodes]),
            "time_to_first_possession_sec": compute_distribution([e.time_to_first_possession_seconds for e in episodes]),
            "possession_changes_per_episode": compute_distribution([float(e.possession_changes) for e in episodes]),

            "passes_attempted_per_episode": compute_distribution([float(e.passes_attempted) for e in episodes]),
            "passes_completed_per_episode": compute_distribution([float(e.passes_completed) for e in episodes]),
            "pass_completion_rate_pct": compute_distribution(
                [e.pass_completion_rate for e in episodes], tot_passes_comp, tot_passes_att
            ),

            "shots_per_episode": compute_distribution([float(e.shots_total) for e in episodes]),
            "shots_on_target_per_episode": compute_distribution([float(e.shots_on_target) for e in episodes]),
            "shot_accuracy_pct": compute_distribution(
                [e.shot_accuracy for e in episodes], tot_shots_on_target, tot_shots
            ),

            "forward_progress_per_episode": compute_distribution([e.forward_progress_total for e in episodes]),
            "min_distance_to_goal": compute_distribution([e.min_ball_dist_to_opp_goal for e in episodes]),
            "max_ball_progress_x": compute_distribution([e.max_ball_progress_x for e in episodes]),

            "tackles_per_episode": compute_distribution([float(e.tackles_attempted) for e in episodes]),
            "interceptions_per_episode": compute_distribution([float(e.interceptions) for e in episodes]),
            "turnovers_conceded_per_episode": compute_distribution([float(e.turnovers_conceded) for e in episodes]),

            "cumulative_reward": compute_distribution([e.cumulative_reward for e in episodes]),
            "action_distribution": action_dist,
            "idle_action_ratio_pct": compute_distribution([e.idle_action_ratio for e in episodes]),
            "pass_ratio_pct": compute_distribution([e.pass_ratio for e in episodes]),
            "shot_ratio_pct": compute_distribution([e.shot_ratio for e in episodes]),
            "dribble_ratio_pct": compute_distribution([e.dribble_ratio for e in episodes]),
            "sprint_ratio_pct": compute_distribution([e.sprint_ratio for e in episodes]),
        }
